### PR TITLE
Ldanzinger/basemapbyreference

### PR DESCRIPTION
**Pull request by ldanzinger**
@JamesMBallard please review/merge. This is 98% a copy of the Generate Offline Map sample. The new additions are a Dialog to ask user whether they want to use a local TPK or generate a new one on the server. If they select to use local, we set the reference file path on the parameters.

### CODE REVIEW

File: ArcGISRuntimeSDKQt_CppSamples/Maps/GenerateOfflineMapLocalBasemap/GenerateOfflineMapLocalBasemap.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        "DownloadButton.qml",
+        "GenerateWindow.qml"
+    ],
+    "name" : "Generate offline map with local basemap",
+    "dataItems" : [
+	    {
+	        "itemId" : "628e8e3521cf45e9a28a12fe10c02c4d",
+	        "path" : "~/ArcGIS/Runtime/Data/tpk/naperville_imagery.tpk"
+	    }
+    ]
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/GenerateOfflineMapLocalBasemap/GenerateOfflineMapLocalBasemap.json]: There are some tabs in this file. Please convert to spaces.

File: ArcGISRuntimeSDKQt_CppSamples/Maps/GenerateOfflineMapLocalBasemap/Info.plist [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	<array>
+		<string>UIInterfaceOrientationPortrait</string>
+		<string>UIInterfaceOrientationPortraitUpsideDown</string>
+		<string>UIInterfaceOrientationLandscapeLeft</string>
+		<string>UIInterfaceOrientationLandscapeRight</string>
+	</array>
+        <key>UILaunchStoryboardName</key>
+        <string>LaunchScreen</string>
+</dict>
+</plist>
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/GenerateOfflineMapLocalBasemap/Info.plist]: Tabs in this file as well. These tend to always contain tabs, so you can leave this one. Bonus points for getting rid of tabs.

File: ArcGISRuntimeSDKQt_CppSamples/Maps/GenerateOfflineMapLocalBasemap/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+There are a number of use-cases where you may wish to use a basemap which is already on the device, rather than downloading:
+
+* You want to limit the total download size.
+* You want to be able to share a single set of basemap files between many offline maps.
+* You want to use a custom basemap (for example authored in ArcGIS Pro) which is not available online.
+* You do not wish to sign into ArcGIS.com in order to download Esri basemaps.
+
+The author of a web map can support the use of basemaps which are already on a device by configuring the web map to specify the name of a suitable basemap file. This could be a basemap which:
+
+* Has been authored in ArcGIS Pro to make use of your organizations custom data.
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/GenerateOfflineMapLocalBasemap/README.md]: `organization's custom data.`

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/GenerateOfflineMapLocalBasemap.qml [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            onButtonClicked: dialog.open()
+        }
+    }
+
+    // Create an extent rectangle for selecting the offline area
+    Rectangle {
+        id: extentRectangle
+        anchors.centerIn: parent
+        width: parent.width - (50)
+        height: parent.height - (125)
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/GenerateOfflineMapLocalBasemap.qml]: I don't see why these numbers need to be parenthesized.

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+There are a number of use-cases where you may wish to use a basemap which is already on the device, rather than downloading:
+
+* You want to limit the total download size.
+* You want to be able to share a single set of basemap files between many offline maps.
+* You want to use a custom basemap (for example authored in ArcGIS Pro) which is not available online.
+* You do not wish to sign into ArcGIS.com in order to download Esri basemaps.
+
+The author of a web map can support the use of basemaps which are already on a device by configuring the web map to specify the name of a suitable basemap file. This could be a basemap which:
+
+* Has been authored in ArcGIS Pro to make use of your organizations custom data.
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/README.md]: `organization's`

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+1. Click on "Generate Offline Map".
+2. You will be prompted to choose whether you wish to download the online basemap or use the "naperville_imagery.tpk" basemap which is already on the device.
+3. If you choose to download the online basemap, the offline map will be generated with the same (topographic) basemap as the online web map.
+4. To download the Esri basemap, you may be prompted to sign in to ArcGIS.com.
+5. If you choose to use the basemap from the device, the offline map will be generated with the local imagery basemap. The download will be quicker since no tiles are exported or downloaded.
+6. Since the application is not exporting online ArcGIS Online basemaps you will not need to log-in.
+
+## How it works
+
+The sample creates a `PortalItem` object using a web map's ID. This portal item is used to initialize an `OfflineMapTask` object. When the button is clicked, the sample requests the default parameters for the task, with the selected extent, by calling `OfflineMapTask::createDefaultGenerateOfflineMapParameters`.
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/README.md]: `OfflineMapTask.createDefaultGenerateOfflineMapParameters` for Qml? I'm not sure if that translates better, but this one is inconsistent with the rest.

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/GenerateOfflineMapLocalBasemap.qml (hunk 4 above)
Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_QMLSamples/Maps/GenerateOfflineMapLocalBasemap/GenerateOfflineMapLocalBasemap.qml]: updated. this is a remnant of our scaleFactor removal find & replace. there are probably a few more instances of these. i cleaned up the 3 generate offline map samples

**Comment by ldanzinger** [comment]
thanks @JamesMBallard - updated

**Pull request merged**